### PR TITLE
HMM2.0

### PR DESCRIPTION
**Pull request by dev-dwarf**
I was hired by @bvisness to work on the list of changes for HMM2.0. The tracking issue for all of the changes is: https://github.com/HandmadeMath/HandmadeMath/issues/147.

## Changes
Here are the changes I made:
- Renaming. Types and functions have been renamed as described in the 2.0 issue: https://github.com/HandmadeMath/HandmadeMath/issues/147. This affects almost every file in the project.
- Angles. A consistent unit is used for angles both as inputs to functions and internally. The default is radians for both, but can be changed. To change the input unit to all HMM functions, define one of `HMM_USE_DEGREE_INPUT` or `HMM_USE_TURN_INPUT`. There are macros, `HMM_AngleRad`, `HMM_AngleDeg`, and `HMM_AngleTurn` that will convert an angle to the input representation, such as passing an FOV in degrees: `HMM_Perspective_RH(HMM_AngleDeg(120), ... other args ...)`. There are also functions to convert angles in the input representation to any of the other units, `HMM_ToRad`, `HMM_ToDeg`, `HMM_ToTurn`.
- Configuration. Added `HMM_PROVIDE_MATH_FUNCTIONS`, removed individual checks for each math function being defined. Users now define this and all of the `HMM_SINF` and other trig functions, or accept the default crt implementations. The `atanf` and `atan2f` functions were unused and have been removed. If users override the functions, they can also use a different unit for the angles in their functions by defining both of `HMM_ANGLE_USER_TO_INTERNAL` and `HMM_ANGLE_INTERNAL_TO_USER`.
- `HMM_PREFIX` macro has been removed. Everything now uses an uppercase HMM_ prefix, so it should be easy to find+replace on your local copy. 
- Handedness. All functions that depended on the handedness of the coordinate system have both a left and right handed version, marked by a "_LH" or "_RH" suffix respectively.
- 2x2 and 3x3 Matrices. Basic operations (Add, Sub, Mul, Div, Transpose) are implemented for each size.
- Matrix Inverses. There are determinants and general inverses for each size of matrix. For 4x4 matrices there are inverses specific to each type of matrix we have a constructor for (Orthographic, Perspective, Rotation, etc). New tests were added for the inverses.
- Overloads and _Generic. There are overloads for the new matrix sizes and the determinant and inverse functions. There is also a C11 _Generic macro for each function that was overloaded before.
- FastNorm is default. The FastNorm functions are now just the Norm functions. Precision has declined in several places from this, but is still on par with the accuracy of GLM. Per: https://github.com/HandmadeMath/HandmadeMath/issues/87
- SLerp. The Slerp implementation was not robust in cases when the quaternions were very close. It has also been changed to always take the shortest path between angles. https://github.com/HandmadeMath/HandmadeMath/issues/135
- `HMM_LinearCombineSSE` is now `HMM_LinearCombineV4M4`. It will use the SSE code when available but fallback to a slower implementation. This reduces the amount of special cases needed for various Mat4 functions, which previously used `HMM_LinearCombineSSE` when available and then each had their own implementation.

## Update Tool
I also created a tool that can be used to update user's code to 2.0 automatically, which you can find in a separate repository here: https://github.com/dev-dwarf/HMM2.0UpdateTool. The tool I created does the following:
- Renames types and functions to use the new naming conventions from: https://github.com/HandmadeMath/HandmadeMath/issues/146
- It wraps arguments that were previously taken in degrees to convert them to radians, per: https://github.com/HandmadeMath/HandmadeMath/pull/107
- Adds the "_RH" suffix to functions that depend on handedness, since now both left-handed and right handed functions are available: https://github.com/HandmadeMath/HandmadeMath/issues/75
- Removes ambiguity with matrix functions now that there are multiple sizes of matrices: https://github.com/HandmadeMath/HandmadeMath/issues/123

The tool is written in one .c file that compiles as C or C++, so any of our users should be able to compile it easily. The tool takes one or more filenames as arguments and updates their contents to 2.0. There are also .bat and .sh scripts to recursively run the tool on all the c/h/cpp/hpp files in a directory. The tool changes the files it touches so it should be used with source control. I can provide compiled executables for windows and linux, but someone else would have to handle mac.

---
Please take a look and let me know anything that needs changes. Sorry for the huge PR, didn't know a good way to split things up. Happy to answer any questions or make additional changes. Thanks for the opportunity!

### CODE REVIEW

**Comment by dev-dwarf** [comment]
Per offline feedback from Ben I've added some commits so that test coverage for the additions is comprehensive. Additionally I've added a configuration option so that projection matrices can use a Z range of [0, 1] for NDCs, this will be helpful for supporting graphics APIs other than OpenGL (DirectX and Metal use this convention that I know of).

**Review by bvisness — CHANGES_REQUESTED** [review]
Looks great overall. I have a few suggested changes but nothing major. I'll test out the renaming script on my Mac and report back with how that goes.

Thanks for all your work on this!

File: HandmadeMath.h [hunk 1]
```diff
@@ -351,43 +368,81 @@ typedef union hmm_vec4
...
 {
     float Elements[4][4];
 
-#ifdef HANDMADE_MATH__USE_SSE
-    __m128 Columns[4];
+    HMM_Vec4 Columns[4];
 
+#ifdef HANDMADE_MATH__USE_SSE
     HMM_DEPRECATED("Our matrices are column-major, so this was named incorrectly. Use Columns instead.")
     __m128 Rows[4];
```
Comment (bvisness) [inline HandmadeMath.h:420]: Let's get rid of this variable. It's been deprecated forever. (Can also just remove `HMM_DEPRECATED`.)

File: HandmadeMath.h [hunk 2]
```diff
@@ -571,18 +640,8 @@ HMM_INLINE float HMM_PREFIX(PowerF)(float Base, float Exponent)
...
-    ASSERT_COVERED(HMM_ToRadians);
-
-    float Result = Degrees * (HMM_PI32 / 180.0f);
-
-    return (Result);
-}
-
 COVERAGE(HMM_Lerp, 1)
-HMM_INLINE float HMM_PREFIX(Lerp)(float A, float Time, float B)
+HMM_INLINE float HMM_Lerp(float A, float Time, float B)
```
Comment (bvisness) [inline HandmadeMath.h:644]: Let's add vector lerp functions - `HMM_LerpV2` etc., plus C++ overloads for `HMM_Lerp`.

**Review by bvisness — CHANGES_REQUESTED** [review]
A couple more things.

File: HandmadeMath.h [hunk 3]
```diff
@@ -1594,13 +2050,29 @@ HMM_INLINE hmm_mat4 HMM_PREFIX(Rotate)(float Angle, hmm_vec3 Axis)
...
+    return HMM_Rotate_RH(-Angle, Axis);
+}
+
+COVERAGE(HMM_InvRotate, 1)
+HMM_INLINE HMM_Mat4 HMM_InvRotate(HMM_Mat4 RotationMatrix)
+{
+    ASSERT_COVERED(HMM_InvRotate);
+
+    return HMM_TransposeM4(RotationMatrix);
+}
```
[the remark this answers is not in the document]
Comment (bvisness) [reply HandmadeMath.h:2068]: Eh whatever; we have `InvTranslate` and `InvScale` so might as well have `InvRotate`.

File: HandmadeMath.h [hunk 4]
```diff
@@ -1507,84 +1470,577 @@ HMM_INLINE hmm_mat4 HMM_PREFIX(DivideMat4f)(hmm_mat4 Matrix, float Scalar)
...
+    HMM_Mat3 Result = Matrix;
 
-    Result.Elements[0][0] = (Axis.X * Axis.X * CosValue) + CosTheta;
-    Result.Elements[0][1] = (Axis.X * Axis.Y * CosValue) + (Axis.Z * SinTheta);
-    Result.Elements[0][2] = (Axis.X * Axis.Z * CosValue) - (Axis.Y * SinTheta);
+    int Columns;
+    for(Columns = 0; Columns < 3; ++Columns)
+    {
+        int Rows;
+        for(Rows = 0; Rows < 3; ++Rows)
```
[the remark this answers is not in the document]
Comment (bvisness) [reply HandmadeMath.h:1553]: Don't worry about this one; #150 will make this complaint obsolete.

**Review by dev-dwarf — COMMENTED** [review]
Hi Ben, please see my comments and let me know if additional changes are required.

**Review by bvisness — APPROVED** [review]
This all looks great. I'm good to merge, but I want to gather feedback from a couple more people first. Will be resolving this in the near future. Thanks again @dev-dwarf!

**Comment by StrangeZak** [comment]
Just took a look through this, and everything looks great. Thanks so much for doing this!!!

**Review by StrangeZak — APPROVED** [review]

**Pull request merged**